python/ray/autoscaler/_private/local/config.py

```diff
@@ -1,2 +1,98 @@
-def bootstrap_local(config):
+import copy
+from typing import Any
+from typing import Dict
+
+from ray.autoscaler._private.cli_logger import cli_logger
+
+unsupported_field_message = ("The field {} is not supported "
+                             "for on-premise clusters.")
+
+
+def prepare_local(config: Dict[str, Any]) -> Dict[str, Any]:
+    """
+    Prepare local cluster config for ingestion by cluster launcher and
+    autoscaler.
+    """
+    config = copy.deepcopy(config)
+    for field in "head_node", "worker_nodes", "available_node_types":
+        if config.get(field):
+            err_msg = unsupported_field_message.format(field)
+            cli_logger.abort(err_msg)
+    if "coordinator_address" in config["provider"]:
+        config = prepare_coordinator(config)
+    else:
+        config = prepare_manual(config)
+    # Presence of head_node and worker_nodes fields triggers legacy config
+    # processing later in the bootstrap process.
+    config["head_node"] = {"placeholder": {}}
+    config["worker_nodes"] = {"placeholder": {}}
+    return config
+
+
+def prepare_coordinator(config: Dict[str, Any]) -> Dict[str, Any]:
+    config = copy.deepcopy(config)
+    # User should explicitly set the max number of workers for the coordinator
+    # to allocate.
+    if "max_workers" not in config:
+        cli_logger.abort("The field `max_workers` is required when using an "
+                         "automatically managed on-premise cluster.")
+    config.setdefault("min_workers", 0)
+    return config
+
+
+def prepare_manual(config: Dict[str, Any]) -> Dict[str, Any]:
+    config = copy.deepcopy(config)
+    if ("worker_ips" not in config["provider"]) or (
+            "head_ip" not in config["provider"]):
+        cli_logger.abort("Please supply a `head_ip` and list of `worker_ips`. "
+                         "Alternatively, supply a `coordinator_address`.")
+    num_ips = len(config["provider"]["worker_ips"])
+    max_specified = "max_workers" in config
+    min_specified = "min_workers" in config
+    if not max_specified and not min_specified:
+        # The most common use-case.
+        config["min_workers"] = num_ips
```

Comment:
is it ?

Reply:
From interactions with users, it seems the main use case for this NodeProvider is to bootstrap Ray at a fixed set of ips. (So, it's not really a cloud provider.) I don't think people are expecting autoscaling to work.

Also, neither the blog post on this nor the local node provider docs mention min_workers or max_workers.

Could modify the defaults for more consistency with the rest of the node providers, but I think it would be more confusing for users.

Reply:
Could be another reason to prefer min_workers=max_workers:
If I read the code right, it seems that downscaling appears to have never worked quite right for this node provider.

When a worker node is deleted, it is simply marked as [terminated](https://github.com/ray-project/ray/blob/6a227ae50132639552b7d3601cc828c28143e2cb/python/ray/autoscaler/_private/local/node_provider.py#L226) in the state file (`ray stop` isn't called). As future (low-priority) work we could try to fix that using, say a node_updater or command_runner in delete_node, but I don't think it's necessary right now.